### ISSUE TITLE
Bad `dict` type inference when assigning `{}` to a dict-like `Protocol` that has `Union` values

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#5552]
**Describe the bug**
pyright seems to make incorrect inferences when assigning an empty dict (`{}`) to a dict-like variable if the type of that variable has methods that return unions as values.

**To Reproduce**
```python
#pyright: strict

from typing import Iterable, Protocol, TypeVar, Tuple

K = TypeVar("K")
V = TypeVar("V")

class SpecialDict(Protocol[K, V]):
    def items(self) -> Iterable[Tuple[K, "V | int"]]: ...
    def __getitem__(self, __key: K) -> "V | int": ...
    def __setitem__(self, __key: K, __value: "V | int") -> None: ...

def foo(k: K, v: V) -> SpecialDict[K, V]:
    # Expression of type "dict[K@foo, V@foo]" cannot be assigned to declared type "SpecialDict[K@foo, V@foo]"
    x: SpecialDict[K, V] = {}
    return x
```

**Expected behavior**
I would expect the inferred type of `{}` to be `Dict[K, V | int]` instead of `Dict[K, V]`, since that is what the `SpecialDict` requires as its protocol signature.

**VS Code extension or command-line**
extension `v1.1.318`

**Additional context**
The example might look contrived in this simplified form, but my actual intention was to make a `NestedDict` protocol, which I don't think can be expressed by using `dict` alone:

```python
class NestedDict(Protocol[K, V]):
    def items(self) -> Iterable[Tuple[K, "V | NestedDict[K, V]"]]: ...
    def __getitem__(self, __key: K) -> "V | NestedDict[K, V]": ...
    def __setitem__(self, __key: K, __value: "V | NestedDict[K, V]") -> None: ...
```

**Comment by FynnBe** [comment]
@Tomaz-Vieira , thank you for looking into this. For reference, I found that a nested dict can be annotated as
```python
K = TypeVar("K")
V = TypeVar("V")
NestedDict = Dict[K, Union["NestedDict[K, V]", V]]
```

**Comment by FynnBe** [comment]
Another option would be to use `collections.UserDict` (didn't manage to properly annotate `dict.items` when inheriting from `dict`)
```python
from collections import UserDict
from typing import ItemsView, TypeVar

K = TypeVar("K")
V = TypeVar("V")


class SpecialDict(UserDict[K, "V | int"]):
    def items(self) -> ItemsView[K, "V | int"]:
        # your special sauce...
        return super().items()


def foo(k: K, v: V) -> SpecialDict[K, V]:
    return SpecialDict({k: v})

```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
